File: backend/app/services/school_service.py
from sqlalchemy.orm import Session
from sqlalchemy import func
from app.models.school import School

def get_all_schools(db: Session):
    """Get all schools with lat/lng from PostGIS"""
    schools = db.query(School).all()
    
    # Convert to dict with lat/lng for frontend
    result = []
    for school in schools:
        lat = db.query(func.ST_Y(School.location)).filter(School.id == school.id).scalar()
        lng = db.query(func.ST_X(School.location)).filter(School.id == school.id).scalar()
        result.append({
            "id": school.id,
            "name": school.name,
            "address": school.address,
            "district": school.district,
            "type": school.type,
            "school_type": str(school.type.value) if hasattr(school.type, 'value') else str(school.type),  # Alias for frontend
            "phone": school.phone,
            "situation": school.situation,
            "email": school.email,
            "score_1": school.score_1,
            "score_2": school.score_2,
            "score_3": school.score_3,
            "score_4": school.score_4,
            "score_5": school.score_5,
            "latitude": lat,
            "longitude": lng,
            "created_at": school.created_at
        })
    return result

def get_nearby_schools(db: Session, latitude: float, longitude: float, radius: float):
    """Find schools within radius using PostGIS"""
    point = func.ST_SetSRID(func.ST_MakePoint(longitude, latitude), 4326)
    
    schools = db.query(School).filter(
        func.ST_DWithin(
            func.ST_Transform(School.location, 3857),
            func.ST_Transform(point, 3857),
            radius
        )
    ).all()
    
    # Convert to dict with lat/lng for frontend
    result = []
    for school in schools:
        lat = db.query(func.ST_Y(School.location)).filter(School.id == school.id).scalar()
        lng = db.query(func.ST_X(School.location)).filter(School.id == school.id).scalar()
        result.append({
            "id": school.id,
            "name": school.name,
            "address": school.address,
            "district": school.district,
            "type": school.type,
            "school_type": str(school.type.value) if hasattr(school.type, 'value') else str(school.type),  # Alias for frontend
            "phone": school.phone,
            "situation": school.situation,
            "email": school.email,
            "score_1": school.score_1,
            "score_2": school.score_2,
            "score_3": school.score_3,
            "score_4": school.score_4,
            "score_5": school.score_5,
            "latitude": lat,
            "longitude": lng,
            "created_at": school.created_at
        })
    return result

File: apps/backend/src/modules/school/school.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { School } from '../../entities/school.entity';

@Injectable()
export class SchoolService {
  constructor(
    @InjectRepository(School) private readonly schoolRepo: Repository<School>,
  ) {}

  private mapRawSchool(raw: Record<string, any>) {
    return {
      id: raw.school_id,
      name: raw.school_name,
      address: raw.school_address,
      district: raw.school_district,
      type: raw.school_type,
      phone: raw.school_phone,
      situation: raw.school_situation,
      email: raw.school_email,
      score_1: raw.school_score_1,
      score_2: raw.school_score_2,
      score_3: raw.school_score_3,
      score_4: raw.school_score_4,
      score_5: raw.school_score_5,
      created_at: raw.school_created_at,
      latitude: parseFloat(raw.latitude),
      longitude: parseFloat(raw.longitude),
    };
  }

  private getBaseQuery() {
    return this.schoolRepo
      .createQueryBuilder('school')
      .select([
        'school.id',
        'school.name',
        'school.address',
        'school.district',
        'school.type',
        'school.phone',
        'school.situation',
        'school.email',
        'school.score_1',
        'school.score_2',
        'school.score_3',
        'school.score_4',
        'school.score_5',
        'school.created_at',
        'ST_Y(school.location::geometry) as latitude',
        'ST_X(school.location::geometry) as longitude',
      ]);
  }

  async getAllSchools() {
    const schools = await this.getBaseQuery().getRawMany();
    return schools.map((s) => this.mapRawSchool(s));
  }

  async getNearbySchools(latitude: number, longitude: number, radius: number) {
    const schools = await this.getBaseQuery()
      .where(
        'ST_DWithin(ST_Transform(school.location::geometry, 3857), ST_Transform(ST_SetSRID(ST_MakePoint(:lng, :lat), 4326), 3857), :radius)',
        { lng: longitude, lat: latitude, radius },
      )
      .getRawMany();
    return schools.map((s) => this.mapRawSchool(s));
  }
}
